### ISSUE TITLE
Latest main build has compiler nullable warning in CI

### ISSUE THREAD

**Issue by Youssef1313** [dotnet/roslyn#81805]
https://dev.azure.com/dnceng-public/public/_build/results?buildId=1239508&view=logs&j=7da62c4d-d31f-5d93-e5c8-c51357236df2&t=a48bff6a-64df-5e1b-ac45-454b273c1d9e

The Source-Build (Managed) job has this warning in CI:

> ##[warning]src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/Extensions/SyntaxGeneratorExtensions_CreateEqualsMethod.cs(243,64): warning CS8604: (NETCORE_ENGINEERING_TELEMETRY=Build) Possible null reference argument for parameter 'fullyQualifiedMetadataName' in 'INamedTypeSymbol? Compilation.GetTypeByMetadataName(string fullyQualifiedMetadataName)'.

Looks like there might be a configuration issue? Warnings should be treated as errors in CI.

**Comment by jjonescz** [comment]
> Warnings should be treated as errors in CI.

Warnings are treated as errors in CI except in the source build legs which I think is by design because those legs use more recent compiler version than others. If this is caused by compiler fixes, I think we will get these same warnings as errors in other legs once we get that compiler version normally through SDK (via an arcade update) and that's when we usually fix them.

But given this doesn't appear in the bootstrap legs, it's probably caused by something else than compiler fixes alone.

**Comment by jjonescz** [comment]
Ah, looks like this diagnostic is suppressed in the normal builds, I see in binlogs there:

<p><samp>
D:\a\_work\1\s\src\Workspaces\SharedUtilitiesAndExtensions\Workspace\Core\Extensions\SyntaxGeneratorExtensions_CreateEqualsMethod.cs(243,64): info SP0001: Diagnostic 'CS8604: Possible null reference argument for parameter 'fullyQualifiedMetadataName' in 'INamedTypeSymbol? Compilation.GetTypeByMetadataName(string fullyQualifiedMetadataName)'.' was programmatically suppressed by a DiagnosticSuppressor with suppression ID 'RS0047' and justification 'The 'Type.FullName' property is never null when using the syntax 'typeof(T).FullName''
</samp></p>

But this suppressor is not included in source build which seems to be by design too:

https://github.com/dotnet/roslyn/blob/9129b60ff204bd192a550ada8b5cb75f9e27819f/eng/targets/Settings.props#L154-L159

That being said, I agree having these "false positive" warnings doesn't seem to be good developer/infra experience, perhaps we can fix that by including that suppressor in source build too if possible.

**Comment by Youssef1313** [comment]
@jjonescz Thanks for the analysis!

Few questions:

1. Why is this suppressor implemented in Roslyn.Diagnostics.Analyzers where it feels like it's more general? Isn't implementing it in Microsoft.CodeAnalysis.NetAnalyzers a more proper place? I personally have seen lots of `typeof(...).FullName!`. Having the suppressor built-in seems like a good idea.
2. Why do you reference Microsoft.CodeAnalysis.NetAnalyzers as a package reference instead of using the analyzer assemblies shipped in SDK?
3. Not an expert on that area, but why all these are excluded on source build? Maybe it's by-design for Microsoft.VisualStudio.Threading.Analyzers if the repo is not enrolled in source-build, but for the others it should be possible to be used during source build? Maybe it's just history back in time before roslyn-analyzers repo enrolled in source-build (https://github.com/dotnet/roslyn-analyzers/pull/5153) ? Now, roslyn-analyzers repo is even fully out of picture as these packages are now in dotnet/roslyn and dotnet/sdk (excluding VS threading analyzers).

**Comment by jjonescz** [comment]
> Why is this suppressor implemented in Roslyn.Diagnostics.Analyzers where it feels like it's more general? Isn't implementing it in Microsoft.CodeAnalysis.NetAnalyzers a more proper place? I personally have seen lots of `typeof(...).FullName!`. Having the suppressor built-in seems like a good idea.

Probably because the former is not officially supported, so the bar for analyzers/suppressors is lower, but it sounds like a good idea to have a feature request on NetAnalyzers to include this.

> Why do you reference Microsoft.CodeAnalysis.NetAnalyzers as a package reference instead of using the analyzer assemblies shipped in SDK?

Good question, I don't know why. Seems like we reference them from roslyn-analyzers too (which doesn't even produce them anymore):

https://github.com/dotnet/roslyn/blob/9129b60ff204bd192a550ada8b5cb75f9e27819f/eng/Version.Details.xml#L146-L147

I will take a look at removing that reference.

> why all these are excluded on source build?

Source build needs to build repos in order. My guess is Roslyn.Diagnostics.Analyzers depends on Roslyn, so it must be built after Roslyn, hence Roslyn cannot reference it.

**Comment by jjonescz** [comment]
Looks like we already have the Roslyn.Diagnostics.Analyzers package available in source build? I can try looking into enabling it there.

https://github.com/dotnet/roslyn/blob/9129b60ff204bd192a550ada8b5cb75f9e27819f/eng/Version.Details.xml#L150-L167

**Comment by jjonescz** [comment]
Apparently this doesn't work in VMR, had to revert it there: https://github.com/dotnet/dotnet/pull/4084/changes/d84689dbbe3221266e3f5bee23cc31be2d1e3db1 due to errors:

```
eng/finish-source-only.proj(137,5): error : (NETCORE_ENGINEERING_TELEMETRY=Build) Prebuilt packages are not allowed in source-only builds.

Detected 2 prebuilt package(s):
  - Microsoft.CodeAnalysis.BannedApiAnalyzers.5.3.0-1.25618.104
  - Microsoft.CodeAnalysis.PublicApiAnalyzers.5.3.0-1.25618.104
```

I guess my initial reasoning stands then - this is not easily possible to fix due to circular dependency.

**Comment by jjonescz** [comment]
Related: https://github.com/dotnet/dotnet/issues/1848
